[HUDI-472] Introduce configurations and new modes of sorting for bulk_insert

yihua opened the pull request:
## What is the purpose of the pull request

Introduce configurations and new modes of sorting for bulk_insert.
- New configuration of sorting for bulk_insert:
  - `hoodie.bulkinsert.sort.mode`: which mode of sorting should be used.  Three sorting modes are implemented: global sort (`"global_sort"`), local sort inside each RDD partition (`"partition_sort"`) and no sort (`"none"`).
- Three modes of bulk_insert regarding sorting:
  - No sort: no sorting is executed, only doing coalesce with repartition disabled in Spark.
  - Global sort: sort the Hudi records across all RDD partitions for bulk_insert, by using `sortBy()` in Spark
  - Local sort inside each RDD partition: only sort the Hudi records within each RDD partition, by using `mapPartitions()` and local sort in Spark

## Brief change log

Major changes inside `hudi-client` package:
- `HoodieWriteConfig`: add the new configuration regarding the sorting mode for bulk_insert
- `CopyOnWriteInsertHandler`, `LazyInsertIterable`, and `BulkInsertMapFunction`: generalize the logic in these classes to handle the steam of Hoodie records which may not be sorted
- `BulkInsertInternalPartitioner`, `GlobalSortPartitioner`, `RDDPartitionSortPartitioner`, `NonSortPartitioner`: implementation of the built-in bulk insert partitioners
- `UserDefinedBulkInsertPartitioner`: extend the interface to indicate whether the records in a RDD partition are sorted
- `BulkInsertHelper`: modify the bulk insert logic to respect the configured sorting mode
- `TestBulkInsertInternalPartitioner`: add unit tests around the new bulk insert partitioners
- `TestCopyOnWriteActionExecutor`: add unit tests to verify the end-to-end behavior of bulk insert in different sorting modes

## Verify this pull request

This change added tests and can be verified as follows:

- New unit tests added in `TestBulkInsertInternalPartitioner` and `TestCopyOnWriteActionExecutor` around the new bulk insert partitioners and the end-to-end behavior of bulk insert in different sorting modes pass.
- All other unit tests in hudi-client pass.

## Committer checklist

 - [ ] Has a corresponding JIRA in PR title & commit
 
 - [ ] Commit message is descriptive of the change
 
 - [ ] CI is green

 - [ ] Necessary doc changes done or have another open PR
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

cdmikechen commented:
I think we can show some improved data size, bulk cost time, sorting cost time and other information using histogram or line chart. In this way, the applicability of each modified method can be better explained.

cdmikechen commented on hudi-client/src/main/java/org/apache/hudi/func/bulkinsert/RDDPartitionLocalSortPartitioner.java:
```diff
@@ -0,0 +1,44 @@
+    return records.mapToPair(record ->
+        new Tuple2<>(
+            String.format("%s+%s", record.getPartitionPath(), record.getRecordKey()), record))
+        .repartitionAndSortWithinPartitions(new HashPartitioner(outputSparkPartitions))
```
`HashPartitioner` means using record key's hashcode value to partition. Will this result in the same range of keys being assigned to different partitions? I think this may cause some problems in subsequent updates.
Maybe `RangePartitioner` is better?

yihua replied:
Yes, `RangePartitioner` is actually what I'm trying to look for.

For this specific Partitioner, it tries to avoid shuffling to speed up the bulk insert but may introduce overlapping ranges.  will check the side effects of this.

yihua commented:
@cdmikechen Thanks for the suggestion.  I'm actually working on benchmarking each mode with different types of workload.  Once I have some data points I'll share them.

vinothchandar commented on hudi-client/src/main/java/org/apache/hudi/func/CopyOnWriteInsertHandler.java:
```diff
@@ -0,0 +1,96 @@
+import org.apache.hudi.table.HoodieTable;
+
+/**
+ * Consumes stream of hoodie records from in-memory queue and writes to one or more create-handles.
```
Improve the javadocs?

yihua replied:
Will do.

vinothchandar commented on hudi-client/src/main/java/org/apache/hudi/func/bulkinsert/BulkInsertInternalPartitioner.java:
```diff
@@ -0,0 +1,38 @@
+import org.apache.hudi.table.UserDefinedBulkInsertPartitioner;
+
+public abstract class BulkInsertInternalPartitioner<T extends HoodieRecordPayload> implements
+    UserDefinedBulkInsertPartitioner<T> {
```
We can rename UserDefinedBulkInsertPartitioner to just `BulkInsertPartitioner` and reuse it I think?

vinothchandar replied:
RDD API clients i.e Uber/marmaray  may need to change the name of the interface implemented. It should be fine IMO. cc @bvaradar @n3nash

yihua replied:
Yeah.  I was also wondering about the custom implementation of `UserDefinedBulkInsertPartitioner` in my previous comment..

n3nash replied:
Can we just keep the UserDefinedBulkInsertPartitioner as an interface and then provide an abstract class here ? Might be fine to change at Uber, but it might be best to avoid a breaking change since it's simple in this case ?

yihua replied:
@vinothchandar @n3nash @bvaradar Let's have a consensus on whether to change the `UserDefinedBulkInsertPartitioner` interface.  I'm fine with either way.  Right now I've also changed the interface so it will be a breaking change.

vinothchandar replied:
I was trying to generalize the existing interface, since what we are doing here is not userdefined anyway.. 

>but it might be best to avoid a breaking change since it's simple in this case ?

this is also fair.. so we can leave the current interface as is and you can just extend/implement it.

vinothchandar commented on hudi-client/src/main/java/org/apache/hudi/func/bulkinsert/NonSortPartitioner.java:
```diff
@@ -0,0 +1,38 @@
+import org.apache.hudi.common.model.HoodieRecordPayload;
+import org.apache.spark.api.java.JavaRDD;
+
+public class NonSortPartitioner<T extends HoodieRecordPayload>
```
More like `NonShufflingPartitioner`?

yihua replied:
Yes, will rename.

n3nash replied:
We are using "Sort" in the other names, should we be consistent ?

vinothchandar commented on hudi-client/src/main/java/org/apache/hudi/table/UserDefinedBulkInsertPartitioner.java:
```diff
@@ -31,4 +31,6 @@
 
   JavaRDD<HoodieRecord<T>> repartitionRecords(JavaRDD<HoodieRecord<T>> records, int outputSparkPartitions);
+
+  boolean arePartitionRecordsSorted();
```
please improve javadocs of this interface while you are here :)

yihua replied:
Yeah, will do.

n3nash replied:
+1

vinothchandar replied:
javadoc please

yihua commented on hudi-client/src/main/java/org/apache/hudi/config/HoodieWriteConfig.java:
```diff
@@ -77,6 +77,10 @@
   private static final String DEFAULT_HOODIE_WRITE_STATUS_CLASS = WriteStatus.class.getName();
   private static final String FINALIZE_WRITE_PARALLELISM = "hoodie.finalize.write.parallelism";
   private static final String DEFAULT_FINALIZE_WRITE_PARALLELISM = DEFAULT_PARALLELISM;
+  private static final String BULKINSERT_SORT_ENABLED = "hoodie.bulkinsert.sort.enable";
```
Yes, one config is better than two for the user to reason.  I'll make the change accordingly.

yihua commented on hudi-client/src/main/java/org/apache/hudi/HoodieWriteClient.java:
```diff
@@ -367,20 +370,30 @@ public static SparkConf registerClasses(SparkConf conf) {
+  }
+
   private JavaRDD<WriteStatus> bulkInsertInternal(JavaRDD<HoodieRecord<T>> dedupedRecords, String commitTime,
       HoodieTable<T> table, Option<UserDefinedBulkInsertPartitioner> bulkInsertPartitioner) {
```
I thought `UserDefinedBulkInsertPartitioner` is user-facing so I didn't touch the interface.  Looking at the source code again, I find that `UserDefinedBulkInsertPartitioner` is only used here.  Then we can just keep a single one as `BulkInsertInternalPartitioner `.

BTW, is this interface (`UserDefinedBulkInsertPartitioner`) intended for user to implement a custom partitioner?  I don't seem to see a config to pass in a custom implementation.  wondering if we should provide such flexibility.

vinothchandar replied:
This was added by @ovj and uber/marmaray project passes in an implementation using the RDD api

n3nash replied:
@yihua Yes, we do need to keep this functionality, at Uber as @vinothchandar pointed out, uses this. We directly pass an implementation rather than a config.

yihua commented:
@vinothchandar Thanks for the code review.  I'll iterate on the code based on the comments and add and fix the unit tests.

n3nash commented on hudi-client/src/main/java/org/apache/hudi/func/bulkinsert/BulkInsertMapFunctionForNonSortedRecords.java:
```diff
@@ -0,0 +1,70 @@
+import org.apache.hudi.func.CopyOnWriteLazyInsertIterable.HoodieInsertValueGenResult;
+import org.apache.hudi.table.HoodieTable;
+
+public class BulkInsertMapFunctionForNonSortedRecords<T extends HoodieRecordPayload>
```
So in this case, we might end up with a skewed Spark Task which ends up writing many files due to large number of records being sent to it (same partition path) ?
@vinothchandar

vinothchandar commented on hudi-client/src/main/java/org/apache/hudi/HoodieWriteClient.java:
```diff
@@ -381,20 +384,30 @@ public static SparkConf registerClasses(SparkConf conf) {
     }
   }
 
+  private BulkInsertMapFunction<T> getBulkInsertMapFunction(
```
I was looking at this for a different JIRA. but having a class for BulkInsertMapfunction seems like an overkill? Can we just replace with lambdas and kill the class ?

eg, in the code on master, just change to below

```
 JavaRDD<WriteStatus> writeStatusRDD = repartitionedRecords
        .mapPartitionsWithIndex((partition, recordItr) ->
            new CopyOnWriteLazyInsertIterable<>(recordItr, config, commitTime, table, fileIDPrefixes.get(partition)), true)
        .flatMap(List::iterator);

```

if you agree, just do it in the PR.. that way I wont introduce an unnecessary rebase for you :)

vinothchandar replied:
cc @yihua

yihua replied:
For some subclasses in this PR, there are more complex logic in `BulkInsertMapFunction`, e.g., `BulkInsertMapFunctionForNonSortedRecords`, wondering if we should keep the API for now?

vinothchandar replied:
Lets do whats easier and gets us moving faster .. :)

vinothchandar commented:
@yihua still on this? 
@n3nash could you review this please?

yihua commented:
@vinothchandar @n3nash I'm still working on resolving one failed unit test.  Somehow all tests pass in CI, but locally I have this single test failed, `org.apache.hudi.table.action.commit.TestCopyOnWriteActionExecutor#testInsertUpsertWithHoodieAvroPayload`:

```
Caused by: java.util.NoSuchElementException: No value present in Option
	at org.apache.hudi.common.util.Option.get(Option.java:88)
	at org.apache.hudi.io.HoodieMergeHandle.<init>(HoodieMergeHandle.java:77)
	at org.apache.hudi.table.action.commit.CommitActionExecutor.getUpdateHandle(CommitActionExecutor.java:118)
	at org.apache.hudi.table.action.commit.CommitActionExecutor.handleUpdate(CommitActionExecutor.java:75)
	at org.apache.hudi.table.action.commit.TestCopyOnWriteActionExecutor.lambda$testInsertUpsertWithHoodieAvroPayload$446d3dad$1(TestCopyOnWriteActionExecutor.java:434)
```

I don't seem to change the merge logic.  If anyone familiar with this test can shed some light, that'd be great.

n3nash commented:
@yihua Is this still failing in the MergeHandle ?

vinothchandar commented:
@yihua is this ready for review?

nsivabalan commented:
@vinothchandar : yes, the patch is ready for review.

yihua commented:
@nsivabalan Thanks for the fix.  There is some ad-hoc code in this PR just for testing.  Let me clean that up.

nsivabalan commented:
sure, thanks. Once done, do ping me and vinoth for review.

vinothchandar commented on hudi-client/src/main/java/org/apache/hudi/config/HoodieWriteConfig.java:
```diff
@@ -245,6 +250,16 @@ public int getMaxConsistencyCheckIntervalMs() {
+    String sortMode = props.getProperty(BULKINSERT_SORT_MODE);
+    try {
+      return BulkInsertSortMode.valueOf(sortMode.toUpperCase());
+    } catch (IllegalArgumentException e) {
```
given IllegalArgumentException is itself runtime exception.. may be ok to just let that percolate.

vinothchandar commented on hudi-client/src/main/java/org/apache/hudi/config/HoodieWriteConfig.java:
```diff
@@ -632,6 +647,10 @@ public FileSystemViewStorageConfig getClientSpecifiedViewStorageConfig() {
   }
 
+  public boolean getStringFormation() {
+    return Boolean.parseBoolean(props.getProperty("hoodie.tmp.string.format"));
```
whats this?

yihua replied:
It's just for local testing to compare `String.format()` vs `StringBuilder`.  `StringBuilder` is more efficient and the reduced overhead is tangible when the number of records is huge.

vinothchandar commented on hudi-client/src/main/java/org/apache/hudi/execution/CopyOnWriteInsertHandler.java:
```diff
@@ -0,0 +1,115 @@
+ * Consumes stream of hoodie records from in-memory queue and writes to one or more create-handles.
+ */
+public class CopyOnWriteInsertHandler<T extends HoodieRecordPayload>
+    extends
```
lets fix the alignment. may be here

vinothchandar commented on hudi-client/src/main/java/org/apache/hudi/execution/CopyOnWriteInsertHandler.java:
```diff
@@ -0,0 +1,109 @@
+
+  @Override
+  public void finish() {
+    for (HoodieWriteHandle handle : handles.values()) {
```
so this implies that `handle.close()` needs to be idempotent? iff we are closing the handle in L85 already, why close everything from the map again?

vinothchandar replied:
okay makes sense. the  key thing seems to be that the handle replaced in the `handles` map, is just closed at the end.

vinothchandar commented on hudi-client/src/main/java/org/apache/hudi/execution/bulkinsert/BulkInsertMapFunctionForNonSortedRecords.java:
```diff
@@ -0,0 +1,76 @@
+public class BulkInsertMapFunctionForNonSortedRecords<T extends HoodieRecordPayload>
+    extends BulkInsertMapFunction<T> {
+
+  Map<String, CopyOnWriteInsertHandler> parallelWritersMap;
```
make this private?

vinothchandar commented on hudi-client/src/main/java/org/apache/hudi/execution/bulkinsert/RDDPartitionRangePartitioner.java:
```diff
@@ -0,0 +1,61 @@
+                .append("+")
+                .append(record.getRecordKey())
+                .toString(), record));
+    Ordering<String> ordering = Ordering$.MODULE$
```
can we write this without importing scala classes here? it will be problematic in terms of bundling..

yihua replied:
I removed the class.

vinothchandar commented on hudi-client/src/main/java/org/apache/hudi/execution/bulkinsert/RDDPartitionRangePartitioner.java:
```diff
@@ -0,0 +1,61 @@
+import scala.reflect.ClassTag;
+import scala.reflect.ClassTag$;
+
+public class RDDPartitionRangePartitioner<T extends HoodieRecordPayload>
```
also please add javadocs for all these classes.. even if its a single line, it makes the code so much more readable in the long run

vinothchandar commented on hudi-client/src/main/java/org/apache/hudi/execution/bulkinsert/RDDPartitionRangePartitioner.java:
```diff
@@ -0,0 +1,61 @@
+                .toString(), record));
+    Ordering<String> ordering = Ordering$.MODULE$
+        .comparatorToOrdering(Comparator.<String>naturalOrder());
+    ClassTag<String> classTag = ClassTag$.MODULE$.apply(String.class);
```
lets implement these using just java class objects ?

yihua replied:
Not relevant anymore.

vinothchandar commented on hudi-client/src/main/java/org/apache/hudi/execution/bulkinsert/RDDPartitionRangePartitioner.java:
```diff
@@ -0,0 +1,61 @@
+    Ordering<String> ordering = Ordering$.MODULE$
+        .comparatorToOrdering(Comparator.<String>naturalOrder());
+    ClassTag<String> classTag = ClassTag$.MODULE$.apply(String.class);
+    return pairRDD.partitionBy(new RangePartitioner<String, HoodieRecord<T>>(
```
is nt this exactly what `sortBy` will do? are we somehow picking a better range? I'd like to skip this implementation otherwise.

vinothchandar commented on hudi-client/src/main/java/org/apache/hudi/table/action/commit/BulkInsertHelper.java:
```diff
@@ -56,31 +59,50 @@
+    boolean arePartitionRecordsSorted = true;
     if (bulkInsertPartitioner.isPresent()) {
-      repartitionedRecords = bulkInsertPartitioner.get().repartitionRecords(dedupedRecords, parallelism);
+      repartitionedRecords = bulkInsertPartitioner.get()
```
just like here.. lets bring 64-65 lines to a single line?

vinothchandar commented on hudi-client/src/main/java/org/apache/hudi/table/action/commit/BulkInsertHelper.java:
```diff
@@ -56,31 +59,50 @@
+      BulkInsertInternalPartitioner partitioner =
+          BulkInsertInternalPartitioner.get(config.getBulkInsertSortMode());
+      repartitionedRecords = partitioner.repartitionRecords(dedupedRecords, parallelism);
+      arePartitionRecordsSorted = partitioner.arePartitionRecordsSorted();
```
this assignment can be done just once outside?  (intellij tips also indicated that IIRC)

vinothchandar commented on hudi-client/src/main/java/org/apache/hudi/table/action/commit/BulkInsertHelper.java:
```diff
@@ -56,31 +59,50 @@
     // generate new file ID prefixes for each output partition
     final List<String> fileIDPrefixes =
-        IntStream.range(0, parallelism).mapToObj(i -> FSUtils.createNewFileIdPfx()).collect(Collectors.toList());
+        IntStream.range(0, parallelism).mapToObj(i -> FSUtils.createNewFileIdPfx())
```
same thing here.  our large monitors to the rescue :) lets make 76-77 into a single line if poss

vinothchandar commented on hudi-client/src/main/java/org/apache/hudi/table/action/commit/BulkInsertHelper.java:
```diff
@@ -56,31 +59,50 @@
         table.getMetaClient().getCommitActionType(), instantTime), Option.empty(),
         config.shouldAllowMultiWriteOnSameInstant());
 
+    /*
```
remove?

vinothchandar commented:
>Do you think we need this? If not, I can remove this class (also getting rid of the scala imports).

@yihua yes lets remove this for now. globally_sorted, local_sorted, no_sort should be good enough for now

yihua commented:
I also simplified the naming of the sorting modes to: `NONE`, `GLOBAL_SORT`, and `PARTITION_SORT`.

vinothchandar commented:
@yihua I see the tests are still failing?

yihua commented:
@vinothchandar @nsivabalan this PR is ready for another review.

I fixed the failing tests.  I also simplified the bulk insert logic regarding different sort modes.  Besides, I added more javadocs and cleaned up the code style.

For the default sort mode for bulk insert, shall we set it to a mode other than GLOBAL_SORT?

vinothchandar commented:
>For the default sort mode for bulk insert, shall we set it to a mode other than GLOBAL_SORT?

yes. sg. we can retain existing behavior.

vinothchandar commented on hudi-common/src/test/java/org/apache/hudi/common/testutils/HoodieTestDataGenerator.java:
```diff
@@ -161,6 +161,17 @@ public static void writePartitionMetadata(FileSystem fs, String[] partitionPaths
     }
   }
 
+  public static List<HoodieRecord> newHoodieRecords(int n, String time) throws Exception {
```
can we not use the existing methods in the data generator to write these tests?

yihua replied:
I copied this method from the `TestCopyOnWriteActionExecutor` class.  Let me check if I can generalize this.

nsivabalan replied:
I see TestRawTripPayload(String jsonData) is being used at many places. Guess we can just move this method to HoodieTestDataGenerator.

yihua replied:
Looks like the method is different from other methods in the generator.  I'll keep it as is.

vinothchandar replied:
but i actually find it surprising we needed to add one. This class as-is is a medley of different individual methods. I would really like to avoid overloading this more.

yihua replied:
After some investigation, I found that tests in `TestCopyOnWriteActionExecutor` use a different record schema, causing the bulk insert to fail using records from `HoodieTestDataGenerator::generateInserts`.  For the bulk insert tests, I use the same schema provided by `HoodieTestDataGenerator` and `generateInserts()` for generating the testing records.  Now the tests work.

nsivabalan commented on hudi-client/src/main/java/org/apache/hudi/execution/bulkinsert/BulkInsertInternalPartitioner.java:
```diff
@@ -0,0 +1,52 @@
+public abstract class BulkInsertInternalPartitioner<T extends HoodieRecordPayload> implements
+    UserDefinedBulkInsertPartitioner<T> {
+
+  public static BulkInsertInternalPartitioner get(BulkInsertSortMode sortMode) {
```
this looks like a factory. did you consider naming with "factory" as suffix or something like that ?

yihua replied:
Yes, good suggestion.  Let me do that.

nsivabalan commented on hudi-client/src/main/java/org/apache/hudi/execution/bulkinsert/RDDPartitionSortPartitioner.java:
```diff
@@ -0,0 +1,68 @@
+          for (; partition.hasNext(); ) {
+            recordList.add(partition.next());
+          }
+          Collections.sort(recordList, (o1, o2) -> o1._1.compareTo(o2._1));
```
will sync up offline with you. interested to know difference between repartition and sort within partitions and current approach. Current approach will bring all records to memory right?

yihua replied:
Yes, let's sync offline.

yihua commented:
I'll take a pass tonight.  Also, @nsivabalan could you `Squash and merge` in Github?  I'd like to keep the commits local in this PR.

vinothchandar commented on hudi-client/src/test/java/org/apache/hudi/execution/bulkinsert/TestBulkInsertInternalPartitioner.java:
```diff
@@ -0,0 +1,143 @@
+  private void verifyRecordAscendingOrder(Iterator<HoodieRecord> records) {
+    HoodieRecord prevRecord = null;
+
+    for (Iterator<HoodieRecord> it = records; it.hasNext(); ) {
```
would n't it be simpler to compare two sorted lists for equivalency? instead of hand-crafting this logic of comparing for ascending order?

yihua replied:
Yes, make sense.  Fixed.

vinothchandar commented on hudi-client/src/test/java/org/apache/hudi/execution/bulkinsert/TestBulkInsertInternalPartitioner.java:
```diff
@@ -0,0 +1,143 @@
+  public static JavaRDD<HoodieRecord> generateTestRecordsForBulkInsert(JavaSparkContext jsc)
+      throws Exception {
+    // RDD partition 1
+    List<HoodieRecord> records1 = newHoodieRecords(3, "2020-07-31T03:16:41.415Z");
```
I think we can write this same test using `dataGen.generateInserts()`?

nsivabalan commented:
I noticed that base interface don't need to be defined as "UserDefined" anymore since its applicable to our own partitioners as well. So, have stripped off "UserDefined" from the naming. Also, have renamed args and variables to userDefinedBulkInsertPartitioner wherever applicable. Once CI succeeds, will land this PR. @yihua : thanks for helping out with this.

yihua commented:
@nsivabalan @n3nash  I just recall that Uber internally extends `UserDefinedBulkInsertPartitioner` to have custom partitioners so the name change of the `UserDefinedBulkInsertPartitioner` class is a breaking change.

Shall we keep the naming `UserDefinedBulkInsertPartitioner`?  We can do this in a separate PR.

nsivabalan commented:
I thought of adding an empty interface called UserDefinedBulkInsertPartitioner, but then BulkInsertPartitioner itself has a new method  boolean arePartitionRecordsSorted(); So, anyways, existing users will have to make code changes to give implementation to this method. So, not sure if there will be benefit adding an empty UserDefinedBulkInsertPartitioner interface extending from BulkInsertPartitioner.

nsivabalan commented:
@n3nash ^

n3nash commented:
@nsivabalan What is the reason to change the name from `UserDefinedBulkInsertPartitioner` to `BulkInsertPartitioner`. I see that you added a new method to the interface, which is fine, but want to understand the reason for the name change

nsivabalan commented:
We have introduced new sort modes as part of this PR and respective bulk insert partitioners. All these were using UserDefinedBulkInsertPartitioner interface, but then these newly added ones are not user defined but thats the only interface we had in place. So it made sense to rename this interface as BulkInsertPartitioner as it is being used by newly added readily available partitioners. But I didn't think too much about existing users. So, open to adding a new empty interface called UserDefinedBulkInsertPartitioner that extends from BulkInsertPartitioner. 
But my point was, even if we do that, we can't avoid users needing to make changes to accommodate the new method. So, my point was why not users implement BulkInsertPartitioner instead of UserDefinedBulkInsertPartitioner incase some changes are required from their end. (BulkInsertPartitioner interface is exactly same as old UserDefinedBulkInsertPartitioner with just an addition of one method arePartitionRecordsSorted() )

nsivabalan commented:
I also thought we could add an abstract class UserDefinedBulkInsertPartitioner and make arePartitionRecordsSorted() false by default, but this also requires users to change from "implements" to "extends" ;)

n3nash commented:
@nsivabalan thanks for the context. I see what you're saying. It's most likely Uber using this special partitioner. I see that a change is required in either scenario, changing the name is fine and keeps it clear without the need to add a placeholder abstract class. We will absorb the backwards incompatible change on our end.
